src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -1570,45 +1539,49 @@ _mongoc_topology_pop_server_session (mongoc_topology_t *topology,
          bson_mutex_unlock (&topology->mutex);
          if (!mongoc_topology_select_server_id (
                 topology, MONGOC_SS_READ, NULL, error)) {
-            RETURN (NULL);
+            ss = NULL;
+            goto done;
          }
 
          bson_mutex_lock (&topology->mutex);
          timeout = td->session_timeout_minutes;
       }
 
       if (timeout == MONGOC_NO_SESSIONS) {
-         bson_mutex_unlock (&topology->mutex);
          bson_set_error (error,
                          MONGOC_ERROR_CLIENT,
                          MONGOC_ERROR_CLIENT_SESSION_FAILURE,
                          "Server does not support sessions");
-         RETURN (NULL);
+         ss = NULL;
+         bson_mutex_unlock (&topology->mutex);
+         goto done;
       }
    }
+   bson_mutex_unlock (&topology->mutex);
 
-   while (topology->session_pool) {
-      ss = topology->session_pool;
-      CDL_DELETE (topology->session_pool, ss);
-      /* Sessions do not expire when the topology type is load balanced. */
-      if (loadbalanced) {
+   while (true) {
+      bool got_session;
+      ss = mongoc_ts_pool_try_pop (topology->session_pool, &got_session);
+      if (!got_session) {
+         /* The pool is empty */
```

Comment:
IIUC, `got_session` may be stale. We may end up creating a few more sessions in the event that many operations are concurrently popping/pushing sessions onto the pool.

Is that correct? If so, that seems like a trade-off to this approach.

Reply:
The `did_pop` bool is somewhat an artefact of an earlier design where you can push/pop arbitrary pointers (including `NULL`), but because of changes to have `alloc_item` and `free_item`, you can never push/pop a `NULL` from the pool, so it is entirely reliable to just check the returned pointer for `NULL` instead of the bool value. I can remove the `did_pop` bool parameter and just check the returned pointer instead.